Photo file name fixes for special characters

chaneylc opened the pull request:
Bugfix for saving photo traits with trait names that have special characters.

Closes #771
Closes #641 
Closes #813 

## Description

_Provide a summary of your changes including motivation and context.
If these changes fix a bug or resolves a feature request, be sure to link to that issue._



## Type of change

_What type of changes does your code introduce? Put an `x` in boxes that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [ ] `FEATURE` (non-breaking change which adds functionality)
- [x] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Checklist:

- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated relevant documentation

## Changelog entry

_Please add a one-line changelog entry below. This will be copied to the changelog file during the release process._

<!-- 
Your release note should be written in clear and straightforward sentences. Most often, users aren't familiar with
the technical details of your PR, so consider what they need to know when you write your release note.

Some brief examples of release notes:
- Fixed a bug causing Field Book to crash when collecting categorical data.
- Added a new option to enable a sound when data is deleted.
- Modified the behavior trait drag and drop behavior.
-->

```release-note
Fixed bug that broke photo saving for trait names with special characters.
```

chaneylc commented:
After digging into this a bit more I realized that the document provider API does automatically sanitize file names when using createFile, so plot data files will automatically work. BUT, createDir does not sanitize file names, so the combination of this and using trait names which could have special characters creates these issues, because they are the names of the directories.

Using the suggestions from https://github.com/PhenoApps/Field-Book/issues/641 I have changed the implementation to make folders based on trait db ids. This change does not require any migration.

chaneylc commented:
- [ ] update audio traits to use trait db id
- [ ] update usb trait to not use plot ids
- [ ] test go pro trait

trife commented:
This isn't currently working. With my test trait, `Test:||Photo`, capturing a photo creates two directories with no images saved.

bellerbrock commented on app/src/main/java/com/fieldbook/tracker/utilities/FileUtil.java:
```diff
@@ -27,6 +27,11 @@ public final class FileUtil {
 
+    //https://stackoverflow.com/questions/2679699/what-characters-allowed-in-file-names-on-android
+    public static String sanitizeFileName(String name) {
+        return name.replaceAll("[|\\?\\*<\"\\\\:>\'\";]*", "_");
```
Hey @chaneylc, just aheads up that this sanitizing strategy will break Breedbase's simple image upload. It assumes photos are named [observationUnitName]_trait_number_timestamp.jpg, so extracts the observationUnitName by capturing everything before the third-to-last underscore.

chaneylc replied:
Ah, I'm assuming breedbases's simple image upload looks like this:
1. collect images with FB
2. export bundle from FB
3. manually upload images into breedbase UI

Exporting the images via brapi should remove the necessity to parse a file name.

Breaking this requirement down: [observationUnitName]_trait_number_timestamp.jpg
1. Are brackets really required around observationUnitName?
2. Obs unit names and trait names often are imported into FB with special characters, Windows and Android file systems do not allow certain characters such as '|'. FB tracks the original name in the database but we have to sanitize these names to make user-friendly-displayable folders for plot data.
3. I noticed that a lot of *Base trait names are very complicated: browsing cassavabase I see
     abiotic stress trait|month 10|COMP:0000054 
My question is, how would you previously import this with the simple breedbase image upload? I can't create a file with this name on my Windows or Android filesystem. 

Would a simple solution for this be changing the breedbase image upload parsing to instead parse the first token after a "_" split?

trife replied:
@bellerbrock @lukasmueller Current thought right now is that we'll sanitize but write the unabridged info to a json in the exif data for each image (#171). That should allow for straightforward parsing but will require a few updates to keep the simple image upload working.

bellerbrock replied:
@chaneylc yep, that's the current workflow. Automating image transfers with brapi would certainly be nice, and would solve this issue.

1. no sorry for the confusion, i just added the brackets to indicate that the observationUnitName is the extraction target

Regarding 2 and 3, my understanding is that photos are traits only within Field-Book to simplify their collection, they aren't defined as formal traits in crop ontology ontologies. By Crop Ontology and Breedbase standards they are akin to nirs spectra, data that is uploaded and associated with specific observation units and then used to derive trait values after analysis. In other words you should not have to worry about naming files with complex crop ontology codes and special characters, the photo trait should be created locally and should exclude them.

A potential solution off the top of my head would be to disallow importing traits of type photo via brapi. And enforce additional restrictions on the trait name (exclude special characters and underscores) when a user creates any trait that will be used to name a file within Field-Book (photo, audio, etc).

Or sticking with the sanitizing strategy you could encode/decode special characters, backslash to escape them, or remove them entirely.  @lukasmueller may have more ideas.

bellerbrock replied:
Ah, missed that comment Trevor. I like the idea of parsing the exif data rather than the filename, that makes a lot more sense

lukasmueller replied:
The standard encoding of special characters in the filesystem is to pre-backslash them. This often works transparently with other software downstream

lukasmueller replied:
but as Trevor pointed out, it doesn't work in Windows :cry:

chaneylc commented:
Added Exif implementation, each image is saved with:
```
/**
         * Json example output saved in TAG_USER_COMMENT EXIF attribute
         *
         * {"study": {
         *  "study_name":"field_sample",
         *  "study_alias":"field_sample",
         *  "study_unique_id_name":"plot_id",
         *  "study_primary_id_name":"row",
         *  "study_secondary_id_name":"plot",
         *  "date_import":"2017-06-15 05:32:44-0700",
         *  "count":200,"internal_id_study":1
         *  },
         *
         * "observation_unit":
         *  {"observation_unit_db_id":"13RPN00002",
         *  "primary_id":"1",
         *  "secondary_id":"2"
         *  },
         *
         *  "observation_variable":{
         *  "observation_variable_name":"Maize Height",
         *  "observation_variable_field_book_format":"photo",
         *  "default_value":"",
         *  "external_db_id":"variable1",
         *  "trait_data_source":"test-server.brapi.org",
         *  "observation_variable_details":"Test"}}
         */
```

Also fixed #803

trife commented:
@chaneylc I think to be all-encompassing, we should include timestamp and collector in the JSON
